File: educative/python/minwindow.py
def minWindow(str, pattern) :
    charsFrequency = {}
    start = 0
    startindex = 0
    minlen = len(str)
    matches = 0
    for char in pattern:
        if char not in charsFrequency:
            charsFrequency[char] = 0
        charsFrequency[char] +=1
    
    for end in range(len(str)):
        rightchar = str[end]
        if rightchar in charsFrequency:
            charsFrequency[rightchar] -=1
            if charsFrequency[rightchar] == 0:
                matches+=1
        while matches == len(pattern):
            if(minlen > end - start +1):
                minlen = end - start  +1
                startindex = start
            leftchar = str[start]
            start +=1
            if leftchar in charsFrequency:
                if charsFrequency[leftchar] == 0:
                    matches -=1
                charsFrequency[leftchar] +=1
    if minlen > len(str):
        return ""
    
    return str[startindex: startindex + minlen]
                
                
            
            
                
            
        
            
            
            
print(minWindow("aabdec", "abc"))

File: educative/cpp/minWindow.cpp
#include <iostream>
#include <string>
#include <unordered_map>
using namespace std;

string minwindow(const string str, const string pattern)
{
  unordered_map<char, int> charFrequencyMap;
  int minlen = str.length();
  int matches = 0;
  int start = 0;
  int startindex = 0;

  for (auto chr : pattern)
  {
    charFrequencyMap[chr]++;
  }
  for (int end = 0; end < str.length(); end++)
  {
    char rightchar = str[end];
    if (charFrequencyMap.find(rightchar) != charFrequencyMap.end())
    {
      charFrequencyMap[rightchar]--;
      if (charFrequencyMap[rightchar] >= 0)
      {
        matches++;
      }
    }
    while (matches == pattern.length())
    {
      if (minlen > end - start + 1)
      {
        minlen = end - start + 1;
        startindex = start;
      }
      char leftchar = str[start];
      start++;
      if (charFrequencyMap.find(leftchar) != charFrequencyMap.end())
      {
        if (charFrequencyMap[leftchar] == 0)
        {
          matches--;
        }
        charFrequencyMap[leftchar]++;
      }
    }
  }

  if (minlen > str.length())
  {
    return "";
  }

  return str.substr(startindex, minlen);
}

int main()
{

  cout << minwindow("aabdec", "abc") << endl;
  cout << minwindow("abdbca", "abc") << endl;
  return 0;
}
